File: src/main/java/net/frozenorb/potpvp/util/EntityUtils.java
package net.frozenorb.potpvp.util;

import org.bukkit.entity.EntityType;

import java.util.EnumMap;
import java.util.Map;

public final class EntityUtils {

    private static Map<EntityType, String> displayNames = new EnumMap<>(EntityType.class);
    private static int currentFakeEntityId = -1;

    public static int getFakeEntityId() {
        return currentFakeEntityId--;
    }

    static {
        displayNames.put(EntityType.ARROW, "Arrow");
        displayNames.put(EntityType.BAT, "Bat");
        displayNames.put(EntityType.BLAZE, "Blaze");
        displayNames.put(EntityType.BOAT, "Boat");
        displayNames.put(EntityType.CAVE_SPIDER, "Cave Spider");
        displayNames.put(EntityType.CHICKEN, "Chicken");
        displayNames.put(EntityType.COMPLEX_PART, "Complex Part");
        displayNames.put(EntityType.COW, "Cow");
        displayNames.put(EntityType.CREEPER, "Creeper");
        displayNames.put(EntityType.DROPPED_ITEM, "Item");
        displayNames.put(EntityType.EGG, "Egg");
        displayNames.put(EntityType.ENDER_CRYSTAL, "Ender Crystal");
        displayNames.put(EntityType.ENDER_DRAGON, "Ender Dragon");
        displayNames.put(EntityType.ENDER_PEARL, "Ender Pearl");
        displayNames.put(EntityType.ENDER_SIGNAL, "Ender Signal");
        displayNames.put(EntityType.ENDERMAN, "Enderman");
        displayNames.put(EntityType.EXPERIENCE_ORB, "Experience Orb");
        displayNames.put(EntityType.FALLING_BLOCK, "Falling Block");
        displayNames.put(EntityType.FIREBALL, "Fireball");
        displayNames.put(EntityType.FIREWORK, "Firework");
        displayNames.put(EntityType.FISHING_HOOK, "Fishing Rod Hook");
        displayNames.put(EntityType.GHAST, "Ghast");
        displayNames.put(EntityType.GIANT, "Giant");
        displayNames.put(EntityType.HORSE, "Horse");
        displayNames.put(EntityType.IRON_GOLEM, "Iron Golem");
        displayNames.put(EntityType.ITEM_FRAME, "Item Frame");
        displayNames.put(EntityType.LEASH_HITCH, "Lead Hitch");
        displayNames.put(EntityType.LIGHTNING, "Lightning");
        displayNames.put(EntityType.MAGMA_CUBE, "Magma Cube");
        displayNames.put(EntityType.MINECART, "Minecart");
        displayNames.put(EntityType.MINECART_CHEST, "Chest Minecart");
        displayNames.put(EntityType.MINECART_FURNACE, "Furnace Minecart");
        displayNames.put(EntityType.MINECART_HOPPER, "Hopper Minecart");
        displayNames.put(EntityType.MINECART_MOB_SPAWNER, "Spawner Minecart");
        displayNames.put(EntityType.MINECART_TNT, "TNT Minecart");
        displayNames.put(EntityType.OCELOT, "Ocelot");
        displayNames.put(EntityType.PAINTING, "Painting");
        displayNames.put(EntityType.PIG, "Pig");
        displayNames.put(EntityType.PIG_ZOMBIE, "Zombie Pigman");
        displayNames.put(EntityType.PLAYER, "Player");
        displayNames.put(EntityType.PRIMED_TNT, "TNT");
        displayNames.put(EntityType.SHEEP, "Sheep");
        displayNames.put(EntityType.SILVERFISH, "Silverfish");
        displayNames.put(EntityType.SKELETON, "Skeleton");
        displayNames.put(EntityType.SLIME, "Slime");
        displayNames.put(EntityType.SMALL_FIREBALL, "Fireball");
        displayNames.put(EntityType.SNOWBALL, "Snowball");
        displayNames.put(EntityType.SNOWMAN, "Snowman");
        displayNames.put(EntityType.SPIDER, "Spider");
        displayNames.put(EntityType.SPLASH_POTION, "Potion");
        displayNames.put(EntityType.SQUID, "Squid");
        displayNames.put(EntityType.THROWN_EXP_BOTTLE, "Experience Bottle");
        displayNames.put(EntityType.UNKNOWN, "Custom");
        displayNames.put(EntityType.VILLAGER, "Villager");
        displayNames.put(EntityType.WEATHER, "Weather");
        displayNames.put(EntityType.WITCH, "Witch");
        displayNames.put(EntityType.WITHER, "Wither");
        displayNames.put(EntityType.WITHER_SKULL, "Wither Skull");
        displayNames.put(EntityType.WOLF, "Wolf");
        displayNames.put(EntityType.ZOMBIE, "Zombie");
    }

    // Static utility class -- cannot be created.
    private EntityUtils() {
    }

    /**
     * Gets the display name of a given entity type.
     *
     * @param type The entity type to lookup.
     * @return The display name of the provided entity type.
     */
    public static String getName(EntityType type) {
        return (displayNames.get(type));
    }

    public static EntityType parse(String input) {
        for (Map.Entry<EntityType, String> entry : displayNames.entrySet()) {
            if (entry.getValue().replace(" ", "").equalsIgnoreCase(input)) {
                return entry.getKey();
            }
        }

        for (EntityType type : EntityType.values()) {
            if (input.equalsIgnoreCase(type.toString())) {
                return type;
            }
        }

        return null;
    }

}

File: src/main/kotlin/net/frozenorb/potpvp/kt/util/EntityUtils.kt
package net.frozenorb.potpvp.kt.util

import org.bukkit.entity.EntityType
import java.util.EnumMap


object EntityUtils {
    private val displayNames: EnumMap<EntityType, String> = EnumMap(EntityType::class.java)
    private var currentFakeEntityId: Int = -1

    @JvmStatic
    fun fakeEntityId(): Int {
        return currentFakeEntityId--
    }

    @JvmStatic
    fun getName(type: EntityType): String? {
        return displayNames[type]
    }

    @JvmStatic
    fun parse(input: String): EntityType? {
        for ((key, value) in displayNames) {
            if (value.replace(" ", "").equals(input, ignoreCase = true)) {
                return key
            }
        }
        for (type in EntityType.values()) {
            if (input.equals(type.toString(), ignoreCase = true)) {
                return type
            }
        }
        return null
    }

    init {
        displayNames[EntityType.ARROW] = "Arrow"
        displayNames[EntityType.BAT] = "Bat"
        displayNames[EntityType.BLAZE] = "Blaze"
        displayNames[EntityType.BOAT] = "Boat"
        displayNames[EntityType.CAVE_SPIDER] = "Cave Spider"
        displayNames[EntityType.CHICKEN] = "Chicken"
        displayNames[EntityType.COMPLEX_PART] = "Complex Part"
        displayNames[EntityType.COW] = "Cow"
        displayNames[EntityType.CREEPER] = "Creeper"
        displayNames[EntityType.DROPPED_ITEM] = "Item"
        displayNames[EntityType.EGG] = "Egg"
        displayNames[EntityType.ENDER_CRYSTAL] = "Ender Crystal"
        displayNames[EntityType.ENDER_DRAGON] = "Ender Dragon"
        displayNames[EntityType.ENDER_PEARL] = "Ender Pearl"
        displayNames[EntityType.ENDER_SIGNAL] = "Ender Signal"
        displayNames[EntityType.ENDERMAN] = "Enderman"
        displayNames[EntityType.EXPERIENCE_ORB] = "Experience Orb"
        displayNames[EntityType.FALLING_BLOCK] = "Falling Block"
        displayNames[EntityType.FIREBALL] = "Fireball"
        displayNames[EntityType.FIREWORK] = "Firework"
        displayNames[EntityType.FISHING_HOOK] = "Fishing Rod Hook"
        displayNames[EntityType.GHAST] = "Ghast"
        displayNames[EntityType.GIANT] = "Giant"
        displayNames[EntityType.HORSE] = "Horse"
        displayNames[EntityType.IRON_GOLEM] = "Iron Golem"
        displayNames[EntityType.ITEM_FRAME] = "Item Frame"
        displayNames[EntityType.LEASH_HITCH] = "Lead Hitch"
        displayNames[EntityType.LIGHTNING] = "Lightning"
        displayNames[EntityType.MAGMA_CUBE] = "Magma Cube"
        displayNames[EntityType.MINECART] = "Minecart"
        displayNames[EntityType.MINECART_CHEST] = "Chest Minecart"
        displayNames[EntityType.MINECART_FURNACE] = "Furnace Minecart"
        displayNames[EntityType.MINECART_HOPPER] = "Hopper Minecart"
        displayNames[EntityType.MINECART_MOB_SPAWNER] = "Spawner Minecart"
        displayNames[EntityType.MINECART_TNT] = "TNT Minecart"
        displayNames[EntityType.OCELOT] = "Ocelot"
        displayNames[EntityType.PAINTING] = "Painting"
        displayNames[EntityType.PIG] = "Pig"
        displayNames[EntityType.PIG_ZOMBIE] = "Zombie Pigman"
        displayNames[EntityType.PLAYER] = "Player"
        displayNames[EntityType.PRIMED_TNT] = "TNT"
        displayNames[EntityType.SHEEP] = "Sheep"
        displayNames[EntityType.SILVERFISH] = "Silverfish"
        displayNames[EntityType.SKELETON] = "Skeleton"
        displayNames[EntityType.SLIME] = "Slime"
        displayNames[EntityType.SMALL_FIREBALL] = "Fireball"
        displayNames[EntityType.SNOWBALL] = "Snowball"
        displayNames[EntityType.SNOWMAN] = "Snowman"
        displayNames[EntityType.SPIDER] = "Spider"
        displayNames[EntityType.SPLASH_POTION] = "Potion"
        displayNames[EntityType.SQUID] = "Squid"
        displayNames[EntityType.THROWN_EXP_BOTTLE] = "Experience Bottle"
        displayNames[EntityType.UNKNOWN] = "Custom"
        displayNames[EntityType.VILLAGER] = "Villager"
        displayNames[EntityType.WEATHER] = "Weather"
        displayNames[EntityType.WITCH] = "Witch"
        displayNames[EntityType.WITHER] = "Wither"
        displayNames[EntityType.WITHER_SKULL] = "Wither Skull"
        displayNames[EntityType.WOLF] = "Wolf"
        displayNames[EntityType.ZOMBIE] = "Zombie"
    }
}